mesonbuild/coredata.py

```diff
@@ -964,15 +964,20 @@ def optionxform(self, optionstr: str) -> str:
         return optionstr
 
 class MachineFileParser():
-    def __init__(self, filenames: T.List[str]) -> None:
+    def __init__(self, filenames: T.List[str], sourcedir: str) -> None:
         self.parser = CmdLineFileParser()
         self.constants: T.Dict[str, T.Union[str, bool, int, T.List[str]]] = {'True': True, 'False': False}
         self.sections: T.Dict[str, T.Dict[str, T.Union[str, bool, int, T.List[str]]]] = {}
 
-        try:
-            self.parser.read(filenames)
-        except configparser.Error as e:
-            raise EnvironmentException(f'Malformed cross or native file: {e}')
+        for f in filenames:
+            with open(f, encoding='utf-8') as fp:
+                content = fp.read()
+                content = content.replace('@GLOBAL_SOURCE_ROOT@', sourcedir)
+                content = content.replace('@DIRNAME@', os.path.dirname(f))
```

Comment:
Could you add some tests for `@@GLOBAL_SOURCE_ROOT@@` and `@@DIRNAME@@`?

Reply:
What should it do? How does it compare to replacements used in custom_targets...

Reply:
What output do you expect, `@GLOBAL_SOURCE_ROOT@` or `@/foo@`? I don't do any escaping here, it's a trivial string replacement, just like in custom_target arguments.

Reply:
Ok. Then no big deal. I was thinking of format strings

Reply:
Added a test with `@/foo@` expected result